GEECS-Scanner-GUI/action_control.py

```diff
@@ -0,0 +1,18 @@
+"""
+Contains a simpler version of a full "RunControl" that is only capable of performing actions using ActionManager
+
+-Chris
+"""
+from geecs_python_api.controls.data_acquisition.data_acquisition import ActionManager
+
+
+class ActionControl:
+    def __init__(self, experiment_name: str):
+        self.action_manager = ActionManager(experiment_dir=experiment_name)
+
+    def perform_action(self, action_list: dict):
+        if 'steps' in action_list and len(action_list['steps']) > 0:
+            self.action_manager.add_action({'closeout_action': action_list})
+            self.action_manager.execute_action('closeout_action')
+
+        # TODO any way to close the TCP connection to the devices we changed?
```

Comment:
ActionManager instantiates it's own set of GeecsDevice objects to interact with. I think I was trying to be "efficient" and so I was having it track the devices it instantiates in an effort to not always instantiate new objects each time it gets interacted with. But, it's true that these objects never close out. I was incorrectly under the impression that TCP subrcribers don't get opened until they are explicitly called for through subscribe_var_vals, but a TCP connection does get opened at initialization of a GeecsDevice. Not sure that is strictly necessary. But, probably the easiest thing to do is to just close the device after each time it gets used in ActionManager. This probably adds a very small amount of overhead that is insignificant in the context of executing real actions.

Reply:
Actually, this comment triggered me to look into a bit more detail of the geecs_device. So, there shouldn't actaully be any tcp_subscribers opened at all by action_manager. Action manager just does basic set/get things via udp, which is simpler communication protocol. So, maybe the question goes back to you a bit, what is it that suggested you wanted to close tcp connections in action manager?

Reply:
Ah ok, if it is UDP and there are no lasting connections and/or it is simpler and easier to maintain then maybe there is no problem at all.  I guess I was looking at the logging statements while testing the "Execute action" buttons and was wondering if these connections do need to be explicitly closed at all.  But from what I could tell there wasn't any active threads when closing the gui so might be all good.